sacred/experiment.py

```diff
@@ -378,24 +389,12 @@ def log_scalar(self, name, value, step=None):
         # The same as Run.log_scalar
         return self.current_run.log_scalar(name, value, step)
 
-    def gather_commands(self):
-        """Iterator over all commands of this experiment.
-
-        Also recursively collects all commands from ingredients.
-
-        Yields
-        ------
-        (str, function)
-            A tuple consisting of the (dotted) command-name and the
-            corresponding captured function.
-
-        """
-        for cmd_name, cmd in self.commands.items():
-            yield cmd_name, cmd
-
-        for ingred in self.ingredients:
-            for cmd_name, cmd in ingred.gather_commands():
-                yield cmd_name, cmd
+    def _gather(self, func):
```

Comment:
I like the idea of a generic gather function a lot. This removes some of the code duplication and as you said, it raises the more descriptive `CircularDependencyError`.

Some minor aesthetics: the usage of this function with the inline lambda expressions is a bit unwieldy. How about using it as a decorator? Something like this:
``` python
@self._gather_from_ingredients
@staticmethod
def gather_commands(ingredient):
    for command_name, command in ingredient.commands.items():
       yield join_paths(ingredient.path, command_name), command
```
Not entirely sure if that works, because of the interaction with `@staticmethod`, but it feels cleaner and more readable. Alternatively having a separate `get_commands` method that is used inside the `gather_commands` method instead of the lambda expression could be good too.

Reply:
I like the idea of the decorator. `self` is unfortunately not defined outside of the function. We can define a decorating function outside of the class body:

```
def gather_from_ingredients(f):
    def wrapped(self):
        for ingredient, _ in self.traverse_ingredients():
            for item in f(ingredient):
                yield item
    return wrapped


class Ingredient(object):
    # ...
    @gather_from_ingredients
    def gather_commands(ingredient):
        for command_name, command in ingredient.commands.items():
            yield join_paths(ingredient.path, command_name), command
```